### PR TITLE
Clean up schema code

### PR DESCRIPTION
**Pull request by kgaillot**

### CODE REVIEW

**Comment by kgaillot** [comment]
@clumens, here's another review when you get a chance

File: lib/common/schemas.c [hunk 1]
```diff
@@ -1053,6 +1053,21 @@ get_schema_version(const char *name)
...
+ *
+ * \param[in,out] xml_blob   XML to update (may be freed and replaced after
+ *                           being transformed)
+ * \param[out]    best       If not NULL, set to schema index of latest schema
+ *                           that validates \p xml_blob
+ * \param[in]     max        If positive, do not update \p xml_blob to any
+ *                           schema past this index
+ * \param[in]     transform  If false, do not update \p xml_blob to any schema
+ *                           that requires an XSL transform
+ * \param[in]     to_logs    Has something to do with logging
```
Comment (clumens) [inline lib/common/schemas.c:1117]: It looks like if this is true, parser errors and warnings will be logged and if it's false, they'll be sent to stderr via fprintf.

File: lib/common/schemas.c [hunk 2]
```diff
@@ -1147,77 +1149,77 @@ update_validation(xmlNode **xml_blob, int *best, int max, gboolean transform,
...
-                          schema->name, lpc, next, max);
-                break;
-            }
+        if (!transform) {
+            /* Validation with this schema succeeded. We are not doing
+             * transforms, so try the next schema using the same XML.
+             */
+            lpc++;
+            continue;
+        }
```
Comment (clumens) [inline lib/common/schemas.c:1209]: I find the style you've switched to here to almost always be much easier to follow.  I'm a big fan of this commit.

File: lib/common/schemas.c [hunk 3]
```diff
@@ -926,8 +926,19 @@ cib_upgrade_err(void *ctx, const char *fmt, ...)
     va_end(ap);
 }
 
+/*!
+ * \internal
+ * \brief Apply a single XSL transformation to given XML
+ *
+ * \param[in] xml        XML to transform
+ * \param[in] transform  XSL name
+ * \param[in] to_logs    Something to do with logging
```
Comment (clumens) [inline lib/common/schemas.c:935]: If true, this causes non-parsing and non-validating error messages to pass through cib_upgrade_err (a function which is just completely out of control) at crm_log_level.

File: lib/common/schemas.c [hunk 4]
```diff
@@ -966,15 +977,26 @@ apply_transformation(xmlNode *xml, const char *transform, gboolean to_logs)
...
  *
- * \note Only emits warnings about enter/leave phases in case of issues.
+ * A schema upgrade can require up to three XSL transformations: an "enter"
+ * transform, the main upgrade transform, and a "leave" transform. Perform
+ * all needed transforms to upgrade given XML to the next schema.
+ *
+ * \param[in] original_xml  XML to transform
+ * \param[in] schema_index  Index of schema that successfully validates
+ *                          \p original_xml
+ * \param[in] to_logs       Something to do with logging
```
Comment (clumens) [inline lib/common/schemas.c:989]: Likewise here.  It's just being passed through to apply_transformation.

File: lib/common/schemas.c (hunk 1 above)
Comment (kgaillot) [reply lib/common/schemas.c:1117]: Yeah for the most part, but there seems to be some other behavior I'm not sure about. I'll put that though :)

Comment (kgaillot) [reply lib/common/schemas.c:1117]: Eventually we'll get rid of all that and use `pcmk__output_t` anyway

**Comment by kgaillot** [comment]
updated pacemaker-next commit to try to fix CI

**Comment by kgaillot** [comment]
added commit on top to update expected test output

**Comment by kgaillot** [comment]
updated commit to fix bug found by coverity

File: lib/common/schemas.c [hunk 5]
```diff
@@ -1053,6 +1058,23 @@ apply_upgrade(const xmlNode *original_xml, int schema_index, gboolean to_logs)
         free(transform_leave);
     }
 
+    if (final == NULL) {
+        return NULL;
+    }
```
Comment (clumens) [inline lib/common/schemas.c:1063]: If you move this up above the previous block, you can also get rid of the `(final != NULL)` part of it.

File: xml/best-match.sh [hunk 6]
```diff
@@ -32,7 +32,7 @@ prefix="$1"; shift
 cd "$(dirname $0)"
 
 list_candidates() {
-    ls -1 "${1}.rng" "${1}"-{[0-9]*,next}.rng 2>/dev/null
+    ls -1 "${1}.rng" "${1}"-[0-9]*.rng "${1}"-next.rng 2>/dev/null
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply xml/best-match.sh:35]: whoops, yep I meant to add this to the first commit

File: lib/common/schemas.c (hunk 5 above)
Comment (kgaillot) [reply lib/common/schemas.c:1063]: final can become NULL in the previous block

**Comment by kgaillot** [comment]
updated to fix commit issue

**Review by clumens — APPROVED** [review]

**Pull request merged**